sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_fetcher.py

```diff
@@ -0,0 +1,199 @@
+from abc import ABC, abstractmethod
+from typing import Dict, Any, List
+
+from azure.cosmos import _retry_utility, http_constants, exceptions
+from azure.cosmos._change_feed.change_feed_start_from import ChangeFeedStartFromPointInTime
+from azure.cosmos._change_feed.change_feed_state import ChangeFeedStateV1, ChangeFeedStateV2
+from azure.cosmos.exceptions import CosmosHttpResponseError
+
+# pylint: disable=protected-access
+
+class ChangeFeedFetcher(ABC):
+
+    @abstractmethod
+    def fetch_next_block(self):
+        pass
+
+class ChangeFeedFetcherV1(ChangeFeedFetcher):
+    """Internal class for change feed fetch v1 implementation.
+     This is used when partition key range id is used or when the supplied continuation token is in just simple etag.
+     Please note v1 does not support split or merge.
+
+    """
+    def __init__(
+            self,
+            client,
+            resource_link: str,
+            feed_options: Dict[str, Any],
+            fetch_function):
+
+        self._client = client
+        self._feed_options = feed_options
+
+        self._change_feed_state: ChangeFeedStateV1 = self._feed_options.pop("changeFeedState")
+        if not isinstance(self._change_feed_state, ChangeFeedStateV1):
+            raise ValueError(f"ChangeFeedFetcherV1 can not handle change feed state version"
+                             f" {type(self._change_feed_state)}")
+
+        self._resource_link = resource_link
+        self._fetch_function = fetch_function
+
+    def fetch_next_block(self):
+        """Returns a block of results.
+
+        :return: List of results.
+        :rtype: list
+        """
+        def callback():
+            return self.fetch_change_feed_items(self._fetch_function)
+
+        return _retry_utility.Execute(self._client, self._client._global_endpoint_manager, callback)
+
+    def fetch_change_feed_items(self, fetch_function) -> List[Dict[str, Any]]:
+        new_options = copy.deepcopy(self._feed_options)
+        new_options["changeFeedState"] = self._change_feed_state
+
+        self._change_feed_state.populate_feed_options(new_options)
+        is_s_time_first_fetch = self._change_feed_state._continuation is None
+        while True:
+            (fetched_items, response_headers) = fetch_function(new_options)
+            continuation_key = http_constants.HttpHeaders.ETag
+            # In change feed queries, the continuation token is always populated. The hasNext() test is whether
+            # there is any items in the response or not.
+            self._change_feed_state.apply_server_response_continuation(
+                response_headers.get(continuation_key))
+
+            if fetched_items:
+                break
+
+            # When processing from point in time, there will be no initial results being returned,
+            # so we will retry with the new continuation token again
+            if (isinstance(self._change_feed_state._change_feed_start_from, ChangeFeedStartFromPointInTime)
```

Comment:
Also here - we should avoid using instance type checks for statements like this, and use discriminators on the objects.

Reply:
updated to use version attribute + ChangeFeedStartFromType enum